File: 2025-03-11/Binary_Tree_Maximum_Path_Sum.java
import java.util.List;

public class Binary_Tree_Maximum_Path_Sum {
    private static int ans;
    public static int solve(TreeNode root) {
        if(root == null) return 0;

        int l = Math.max(0, solve(root.left));
        int r = Math.max(0, solve(root.right));

        ans = Math.max(ans, root.val+l+r);

        return root.val + Math.max(l, r);
    }
    public static int maxPathSum(TreeNode root) {
        ans = root.val;
        solve(root);
        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{1,-2,3};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        int ans = maxPathSum(root);
        System.out.println(ans);
    }
}

File: 2025-10-01/Binary_Tree_Maximum_Path_Sum.py
from collections import deque
from typing import Optional, List

from TreeNode import TreeNode
from Tree import Tree


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return True

        ans = [root.val]
        def dfs(root):
            if not root:
                return 0

            l = max(0, dfs(root.left))
            r = max(0, dfs(root.right))

            ans[0] = max(ans[0], root.val + l + r)
            return root.val + max(l, r)

        dfs(root)
        return ans[0]


sol = Solution()
tree = Tree()
node = tree.array_to_tree([-10,9,20,None,None,15,7])
print(sol.maxPathSum(node))
